File: tools/ingestion/legisnote_ingest/parse/czech_statute.py
"""Deterministic parser: Czech statute plain text / Markdown -> IR unit tree.

Recognises the standard Czech statutory hierarchy (docs/research-czech-legislation-data.md §6):

    Část (part)            ->  ČÁST PRVNÍ
    Hlava (title)          ->  HLAVA I
    Oddíl / Díl (chapter)  ->  Oddíl 1
    § (section)            ->  § 12   [optional heading on the next line]
    odstavec (paragraph)   ->  (1), (2)
    písmeno / bod (point)  ->  a), b) / 1.

This is the shared normalization step reused by every text-bearing adapter
(LawGPT Markdown, zakonyprolidi text, PDF-extracted text). It is a v1 heuristic
parser — good enough for the PoC (91/2012 Sb.) and unit-tested; refine as more
laws are ingested. Truly structured JSON sources (eSbírka) bypass this and map
fragments to IR directly.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field

from ..ir import NodeType, Unit

# Hierarchy depth per node type (smaller = higher in the tree).
_DEPTH: dict[NodeType, int] = {
    "part": 1,
    "title": 2,
    "chapter": 3,
    "section": 4,
    "paragraph": 5,
    "point": 6,
}

_PART = re.compile(r"^#*\s*ČÁST\s+(.+)", re.IGNORECASE)
_TITLE = re.compile(r"^#*\s*HLAVA\s+(.+)", re.IGNORECASE)
_CHAPTER = re.compile(r"^#*\s*(?:Oddíl|Díl|Kapitola)\s+(.+)", re.IGNORECASE)
_SECTION = re.compile(r"^#*\s*§\s*(\d+[a-z]?)\b(.*)$")
_PARAGRAPH = re.compile(r"^\(\s*(\d+)\s*\)\s*(.*)$")
_POINT = re.compile(r"^([a-zA-Zľščřžýáíéúůňťďóě])\)\s*(.*)$")


@dataclass
class _Node:
    node_type: NodeType
    token: str  # key fragment, e.g. "s12", "o2", "pa", "cast1"
    label: str | None
    ordinal: int
    text_lines: list[str] = field(default_factory=list)
    children: list["_Node"] = field(default_factory=list)
    depth: int = 0

    def to_unit(self, key_prefix: str) -> Unit:
        node_key = f"{key_prefix}/{self.token}" if key_prefix else self.token
        text = "\n".join(self.text_lines).strip() or None
        return Unit(
            node_key=node_key,
            node_type=self.node_type,
            label=self.label,
            ordinal=self.ordinal,
            text=text,
            children=[c.to_unit(node_key) for c in self.children],
        )


def parse_czech_statute(text: str) -> list[Unit]:
    """Parse statute text into a tree of IR :class:`Unit` roots."""
    roots: list[_Node] = []
    stack: list[_Node] = []

    def _parent_for(depth: int) -> _Node | None:
        while stack and stack[-1].depth >= depth:
            stack.pop()
        return stack[-1] if stack else None

    def _attach(node: _Node) -> None:
        parent = _parent_for(node.depth)
        siblings = parent.children if parent else roots
        node.ordinal = len(siblings)
        siblings.append(node)
        stack.append(node)

    for raw in text.splitlines():
        line = raw.strip()
        if not line:
            continue

        if m := _PART.match(line):
            n = _Node("part", "", line, 0, depth=_DEPTH["part"])
            n.token = f"cast{_count(roots, 'part') + 1}"
            _attach(n)
        elif m := _TITLE.match(line):
            n = _Node("title", "", line, 0, depth=_DEPTH["title"])
            _attach(n)
            n.token = f"hlava{n.ordinal + 1}"
        elif m := _CHAPTER.match(line):
            n = _Node("chapter", "", line, 0, depth=_DEPTH["chapter"])
            _attach(n)
            n.token = f"oddil{n.ordinal + 1}"
        elif m := _SECTION.match(line):
            num = m.group(1)
            inline = m.group(2).strip()
            n = _Node("section", f"s{num}", f"§ {num}", 0, depth=_DEPTH["section"])
            if inline:
                n.text_lines.append(inline)
            _attach(n)
        elif m := _PARAGRAPH.match(line):
            num = m.group(1)
            rest = m.group(2)
            n = _Node("paragraph", f"o{num}", f"({num})", 0, depth=_DEPTH["paragraph"])
            if rest:
                n.text_lines.append(rest)
            _attach(n)
        elif m := _POINT.match(line):
            letter = m.group(1).lower()
            rest = m.group(2)
            n = _Node("point", f"p{letter}", f"{letter})", 0, depth=_DEPTH["point"])
            if rest:
                n.text_lines.append(rest)
            _attach(n)
        else:
            # Continuation text — belongs to the deepest open unit.
            if stack:
                stack[-1].text_lines.append(line)
            # Text before any structural marker (preamble) is dropped from the
            # body tree on purpose; the importer takes law title/preamble from
            # the manifest's `law` block.

    return [r.to_unit("") for r in roots]


def _count(nodes: list[_Node], node_type: NodeType) -> int:
    return sum(1 for n in nodes if n.node_type == node_type)

File: apps/web/src/server/import/czechStatute.ts
import type { ManifestUnit, NodeType } from "@legisnote/shared";

/**
 * Deterministic Czech-statute parser (TS port of tools/ingestion's
 * parse/czech_statute.py) so the web app can import LawGPT Markdown directly.
 * Recognises ČÁST / HLAVA / Oddíl·Díl·Kapitola / § / (n) odstavec / a) písmeno
 * and assigns stable nodeKeys (cast1/s1/o2/pa) that survive renumbering (FR-10a).
 * Heuristic v1 — keep in sync with the Python parser as laws are added.
 */

const DEPTH: Record<string, number> = {
  part: 1,
  title: 2,
  chapter: 3,
  section: 4,
  paragraph: 5,
  point: 6,
};

const RE_PART = /^#*\s*ČÁST\s+(.+)/i;
const RE_TITLE = /^#*\s*HLAVA\s+(.+)/i;
const RE_CHAPTER = /^#*\s*(?:Oddíl|Díl|Kapitola)\s+(.+)/i;
const RE_SECTION = /^#*\s*§\s*(\d+[a-z]?)\b(.*)$/;
const RE_PARAGRAPH = /^\(\s*(\d+)\s*\)\s*(.*)$/;
const RE_POINT = /^([a-zA-Zľščřžýáíéúůňťďóě])\)\s*(.*)$/;

interface Node {
  nodeType: NodeType;
  token: string;
  label: string | null;
  ordinal: number;
  textLines: string[];
  children: Node[];
  depth: number;
}

function newNode(nodeType: NodeType, token: string, label: string | null, depth: number): Node {
  return { nodeType, token, label, ordinal: 0, textLines: [], children: [], depth };
}

function toUnit(node: Node, keyPrefix: string): ManifestUnit {
  const nodeKey = keyPrefix ? `${keyPrefix}/${node.token}` : node.token;
  const text = node.textLines.join("\n").trim() || null;
  return {
    nodeKey,
    nodeType: node.nodeType,
    label: node.label,
    ordinal: node.ordinal,
    text,
    children: node.children.map((c) => toUnit(c, nodeKey)),
  };
}

export function parseCzechStatute(text: string): ManifestUnit[] {
  const roots: Node[] = [];
  const stack: Node[] = [];

  const parentFor = (depth: number): Node | null => {
    while (stack.length && stack[stack.length - 1]!.depth >= depth) stack.pop();
    return stack.length ? stack[stack.length - 1]! : null;
  };

  const attach = (node: Node): void => {
    const parent = parentFor(node.depth);
    const siblings = parent ? parent.children : roots;
    node.ordinal = siblings.length;
    siblings.push(node);
    stack.push(node);
  };

  const countType = (nodes: Node[], t: NodeType): number => nodes.filter((n) => n.nodeType === t).length;

  for (const raw of text.split(/\r?\n/)) {
    const line = raw.trim();
    if (!line) continue;

    let m: RegExpMatchArray | null;
    if (RE_PART.test(line)) {
      const n = newNode("part", "", line, DEPTH.part!);
      n.token = `cast${countType(roots, "part") + 1}`;
      attach(n);
    } else if (RE_TITLE.test(line)) {
      const n = newNode("title", "", line, DEPTH.title!);
      attach(n);
      n.token = `hlava${n.ordinal + 1}`;
    } else if (RE_CHAPTER.test(line)) {
      const n = newNode("chapter", "", line, DEPTH.chapter!);
      attach(n);
      n.token = `oddil${n.ordinal + 1}`;
    } else if ((m = line.match(RE_SECTION))) {
      const num = m[1]!;
      const inline = (m[2] ?? "").trim();
      const n = newNode("section", `s${num}`, `§ ${num}`, DEPTH.section!);
      if (inline) n.textLines.push(inline);
      attach(n);
    } else if ((m = line.match(RE_PARAGRAPH))) {
      const num = m[1]!;
      const rest = m[2] ?? "";
      const n = newNode("paragraph", `o${num}`, `(${num})`, DEPTH.paragraph!);
      if (rest) n.textLines.push(rest);
      attach(n);
    } else if ((m = line.match(RE_POINT))) {
      const letter = m[1]!.toLowerCase();
      const rest = m[2] ?? "";
      const n = newNode("point", `p${letter}`, `${letter})`, DEPTH.point!);
      if (rest) n.textLines.push(rest);
      attach(n);
    } else if (stack.length) {
      // Continuation text belongs to the deepest open unit. Preamble before any
      // structural marker is dropped on purpose (title comes from the manifest).
      stack[stack.length - 1]!.textLines.push(line);
    }
  }

  return roots.map((r) => toUnit(r, ""));
}
